homeassistant/components/template/vacuum.py

```diff
@@ -0,0 +1,364 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_VACUUMS = 'vacuums'
+CONF_BATTERY_LEVEL_TEMPLATE = 'battery_level_template'
+CONF_FAN_SPEED_LIST = 'fan_speeds'
+CONF_FAN_SPEED_TEMPLATE = 'fan_speed_template'
+
+ENTITY_ID_FORMAT = DOMAIN + '.{}'
+_VALID_STATES = [STATE_CLEANING, STATE_DOCKED, STATE_PAUSED, STATE_IDLE,
+                 STATE_RETURNING, STATE_ERROR]
+
+VACUUM_SCHEMA = vol.Schema({
+    vol.Optional(CONF_FRIENDLY_NAME): cv.string,
+    vol.Optional(CONF_VALUE_TEMPLATE): cv.template,
+    vol.Optional(CONF_BATTERY_LEVEL_TEMPLATE): cv.template,
+    vol.Optional(CONF_FAN_SPEED_TEMPLATE): cv.template,
+
+    vol.Required(SERVICE_START): cv.SCRIPT_SCHEMA,
+    vol.Optional(SERVICE_PAUSE): cv.SCRIPT_SCHEMA,
+    vol.Optional(SERVICE_STOP): cv.SCRIPT_SCHEMA,
+    vol.Optional(SERVICE_RETURN_TO_BASE): cv.SCRIPT_SCHEMA,
+    vol.Optional(SERVICE_CLEAN_SPOT): cv.SCRIPT_SCHEMA,
+    vol.Optional(SERVICE_LOCATE): cv.SCRIPT_SCHEMA,
+    vol.Optional(SERVICE_SET_FAN_SPEED): cv.SCRIPT_SCHEMA,
+
+    vol.Optional(
+        CONF_FAN_SPEED_LIST,
+        default=[]
+    ): cv.ensure_list,
+
+    vol.Optional(CONF_ENTITY_ID): cv.entity_ids
+})
+
+PLATFORM_SCHEMA = cv.PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_VACUUMS): vol.Schema({cv.slug: VACUUM_SCHEMA}),
+})
+
+
+async def async_setup_platform(
+        hass, config, async_add_entities, discovery_info=None
+):
+    """Set up the Template Vacuums."""
+    vacuums = []
+
+    for device, device_config in config[CONF_VACUUMS].items():
+        friendly_name = device_config.get(CONF_FRIENDLY_NAME, device)
+
+        state_template = device_config.get(CONF_VALUE_TEMPLATE)
+        battery_level_template = device_config.get(CONF_BATTERY_LEVEL_TEMPLATE)
+        fan_speed_template = device_config.get(CONF_FAN_SPEED_TEMPLATE)
+
+        start_action = device_config[SERVICE_START]
+        pause_action = device_config.get(SERVICE_PAUSE)
+        stop_action = device_config.get(SERVICE_STOP)
+        return_to_base_action = device_config.get(SERVICE_RETURN_TO_BASE)
+        clean_spot_action = device_config.get(SERVICE_CLEAN_SPOT)
+        locate_action = device_config.get(SERVICE_LOCATE)
+        set_fan_speed_action = device_config.get(SERVICE_SET_FAN_SPEED)
+
+        fan_speed_list = device_config[CONF_FAN_SPEED_LIST]
+
+        entity_ids = set()
+        manual_entity_ids = device_config.get(CONF_ENTITY_ID)
+        invalid_templates = []
+
+        for tpl_name, template in (
+                (CONF_VALUE_TEMPLATE, state_template),
+                (CONF_BATTERY_LEVEL_TEMPLATE, battery_level_template),
+                (CONF_FAN_SPEED_TEMPLATE, fan_speed_template)
+        ):
+            if template is None:
+                continue
+            template.hass = hass
+
+            if manual_entity_ids is not None:
+                continue
+
+            template_entity_ids = template.extract_entities()
+            if template_entity_ids == MATCH_ALL:
+                entity_ids = MATCH_ALL
```

Comment:
This means that all templates for a device that are defined must have entities defined, otherwise we match all. Probably makes sense.

Reply:
This change was based on the template sensor (see [previous PR](https://github.com/home-assistant/home-assistant/pull/17215#pullrequestreview-200423988)), but I just noticed I missed the [check before starting state change tracking](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/template/sensor.py#L143-L146). I'll also add that check for the template vacuum.